sdk/identity/identity/src/credentials/authorizationCodeCredential.ts

```diff
@@ -0,0 +1,145 @@
+   * from Azure Active Directory.
+   *
+   * It is currently necessary for the user of this credential to initiate
+   * the authorization code flow to obtain an authorization code to be used
+   * with this credential.  A full example of this flow is provided here:
+   *
+   * https://github.com/daviwil/azure-sdk-for-js/blob/master/sdk/identity/identity/samples/authorizationCodeSample.ts
+   *
+   * @param tenantId The Azure Active Directory tenant (directory) ID or name.
+   * @param clientId The client (application) ID of an App Registration in the tenant.
+   * @param clientSecret A client secret that was generated for the App Registration or
+                         'undefined' if using this credential in a desktop or mobile
+                         application.
+   * @param authorizationCode An authorization code that was received from following the
+                              authorization code flow.  This authorization code must not
+                              have already been used to obtain an access token.
+   * @param redirectUri The redirect URI that was used to request the authorization code.
+                        Must be the same URI that is configured for the App Registration.
+   * @param options Options for configuring the client which makes the access token request.
+   */
+  constructor(
+    tenantId: string,
+    clientId: string,
+    clientSecret: string | undefined,
+    authorizationCode: string,
+    redirectUri: string,
+    options?: IdentityClientOptions
+  ) {
+    this.identityClient = new IdentityClient(options);
+    this.tenantId = tenantId;
+    this.clientId = clientId;
+    this.clientSecret = clientSecret;
+    this.authorizationCode = authorizationCode;
+    this.redirectUri = redirectUri;
+  }
+
+  /**
+   * Authenticates with Azure Active Directory and returns an {@link AccessToken} if
+   * successful.  If authentication cannot be performed at this time, this method may
+   * return null.  If an error occurs during authentication, an {@link AuthenticationError}
+   * containing failure details will be thrown.
+   *
+   * @param scopes The list of scopes for which the token will have access.
+   * @param options The options used to configure any requests this
+   *                TokenCredential implementation might make.
+   */
+  public async getToken(
+    scopes: string | string[],
+    options?: GetTokenOptions
+  ): Promise<AccessToken | null> {
+    const { span, options: newOptions } = createSpan(
+      "AuthorizationCodeCredential-getToken",
+      options
+    );
+    try {
+      let tokenResponse: TokenResponse | null = null;
+      let scopeString = typeof scopes === "string" ? scopes : scopes.join(" ");
+      if (scopeString.indexOf("offline_access") < 0) {
+        scopeString += " offline_access";
+      }
+
+      // Try to use the refresh token first
+      if (this.lastTokenResponse && this.lastTokenResponse.refreshToken) {
```

Comment:
This implies `getToken` should be called only when a new access token is wanted, i.e. the caller is responsible for any caching?

Reply:
This is the correct behavior.  Without the aid of MSAL in node.js the best behavior is to cache the token next to where it is used, so the credential types don't need to maintain a scope based cache of access tokens.

Reply:
Yep, we cache the token in the HTTP pipeline's authentication policy.  The token expiration is handled there and `getToken` will only be called after the authentication policy expects that the token will expire soon.